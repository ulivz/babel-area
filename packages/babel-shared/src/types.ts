import * as t from '@babel/types';
import { Visitor, BabelFile, PluginOptions } from '@babel/core';

export { t, PluginOptions };

export interface PluginPass<T extends PluginOptions> {
  file: BabelFile;
  key: string;
  opts: T;
  cwd: string;
  filename: string;
  [key: string]: unknown;
}

export interface PluginObj<T extends PluginOptions, U extends object = object, S = PluginPass<T> & U> {
  name?: string;
  manipulateOptions?(opts: T, parserOpts: any): void;
  pre?(this: S, file: BabelFile): void;
  visitor: Visitor<S>;
  post?(this: S, file: BabelFile): void;
  inherits?: any;
}

/**
 * Expose type of babel-plugin
 */
export type BabelPlugin<T extends PluginOptions = object, U extends object = object> = (
  babel: { types: typeof t; }
) => PluginObj<T, U>

declare module '@babel/traverse' {
  export interface Hub {
    file: {
      metadata: { [key: string]: any }
    }
  }
  // path.scope.getProgramParent().data.imports
  // @ref: https://github.com/babel/babel/blob/1945bafcf3599292f7ebeccd5c7c2635dd1cb0d5/packages/babel-traverse/src/scope/index.ts#L392
  interface Scope {
    data: { [key: string | symbol]: unknown };
  }
}
