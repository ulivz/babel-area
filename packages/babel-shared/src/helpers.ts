import * as t from '@babel/types';
import { BabelPlugin, PluginOptions } from './types';

export function declarePlugin<
  T extends PluginOptions = object,
  U extends object = object
>(fn: BabelPlugin<T, U>): typeof fn {
  return fn;
}

export function declarePluginTuple<T extends PluginOptions = object, U extends object = object>(
  plugin: BabelPlugin<T, U>,
  pluginOptions?: T,
): [BabelPlugin<T, U>, T] {
  return [plugin, pluginOptions || {} as T];
}

/**
 * Expose `buildMemberExpressionByIdentifierHierarchy`.
 */
export function buildMemberExpressionByIdentifierHierarchy(
  hierarchy: Array<string | t.Identifier | t.ThisExpression>,
) {
  const normalizedHierarchy = hierarchy.map(
    path => (typeof path === 'string' ? t.identifier(path) : path),
  );

  let currentMemberExpression: t.MemberExpression;

  for (let i = 1; i <= normalizedHierarchy.length - 1; i++) {
    const left = normalizedHierarchy[i - 1];
    const right = normalizedHierarchy[i];
    if (left && right) {
      currentMemberExpression = t.memberExpression(
        i === 1 ? left : currentMemberExpression!,
        right,
      );
    }
  }

  return currentMemberExpression!;
}
