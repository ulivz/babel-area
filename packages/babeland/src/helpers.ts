import { codeFrameColumns, SourceLocation, BabelCodeFrameOptions } from '@babel/code-frame';
import { Visitor } from '@babel/traverse';

export function defineVisitor<T extends object = object>(visitor: Visitor<T>): Visitor<T> {
  return visitor;
}

/**
 * Expose `HighlightCodeError`
 */
export class HighlightCodeError extends Error {
  constructor(
    source: string,
    location: SourceLocation,
    message: string,
    options?: BabelCodeFrameOptions,
  ) {
    super(message);
    this.message = getHighlightCodeString(
      source,
      location,
      message,
      options,
    );
  }
}

/**
 * Expose `getHighlightCodeString`
 */
export function getHighlightCodeString(
  source: string,
  location: SourceLocation,
  message: string,
  options?: BabelCodeFrameOptions,
  correctColumn = true,
): string {
  if (correctColumn && location?.start?.column) {
    location.start.column += 1;
  }

  return codeFrameColumns(
    source,
    location,
    {
      highlightCode: true,
      forceColor: true,
      message,
      linesAbove: 10,
      linesBelow: 2,
      ...(options || {}),
    },
  );
}
