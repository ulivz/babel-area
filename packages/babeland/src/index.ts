import { parse } from '@babel/parser';
import generate from '@babel/generator';
import traverse, { VisitNode, NodePath, Binding } from '@babel/traverse';

export * from './transform';
export * from './helpers';
export {
  transformSync,
  transformAsync,
  transformFromAstSync,
  transformFromAstAsync,
} from '@babel/core';
export * from 'babel-shared';
export {
  parse,
  traverse,
  generate,
  VisitNode,
  NodePath,
  Binding,
};
