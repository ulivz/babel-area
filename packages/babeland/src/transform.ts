

/**
  * FIXME: how can we make that babel does not format code
  */
export const DEFAULT_TRANSFORM_OPTIONS: babel.TransformOptions = {
  /**
    * do not enable config file
    */
  configFile: false,
  /**
    * do not read babelrc
    */
  babelrc: false,
  compact: false,
  /**
    * enable typescript
    */
  parserOpts: {
    sourceType: 'module',
    plugins: [
      'typescript',
    ],
  },
};
