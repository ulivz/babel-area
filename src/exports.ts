/**
 * Module dependencies
 */
import { parse } from '@babel/parser';
import generate from '@babel/generator';
import traverse, { VisitNode, NodePath, Binding } from '@babel/traverse';

export * from './transform';
export * from './types';
export * from './helpers';

export {
  parse,
  traverse,
  generate,
  VisitNode,
  NodePath,
  Binding,
};
