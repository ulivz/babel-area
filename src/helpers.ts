/**
 * Module dependencies
 */
import * as t from '@babel/types';
import { codeFrameColumns, SourceLocation, BabelCodeFrameOptions } from '@babel/code-frame';
import { Visitor } from '@babel/traverse';
import { BabelPlugin, PluginOptions } from './types';

export function declarePlugin<
  T extends PluginOptions = object,
  U extends object = object
>(fn: BabelPlugin<T, U>): typeof fn {
  return fn;
}

export function defineVisitor<T extends object = object>(visitor: Visitor<T>): Visitor<T> {
  return visitor;
}

export function declarePluginConfig<T extends PluginOptions = object>(
  plugin: BabelPlugin<T>,
  pluginOptions?: T,
): [BabelPlugin<T>, T] {
  return [plugin, pluginOptions || {} as T];
}

/**
 * Expose `getHighlightCodeString`
 */
export function getHighlightCodeString(
  source: string,
  location: SourceLocation,
  message: string,
  options?: BabelCodeFrameOptions,
  correctColumn = true,
): string {
  if (correctColumn && location?.start?.column) {
    location.start.column += 1;
  }

  return codeFrameColumns(
    source,
    location,
    {
      highlightCode: true,
      forceColor: true,
      message,
      linesAbove: 10,
      linesBelow: 2,
      ...(options || {}),
    },
  );
}

/**
 * Expose `HighlightCodeError`
 */
export class HighlightCodeError extends Error {
  constructor(
    source: string,
    location: SourceLocation,
    message: string,
    options?: BabelCodeFrameOptions,
  ) {
    super(message);
    this.message = getHighlightCodeString(
      source,
      location,
      message,
      options,
    );
  }
}

/**
 * Expose `buildMemberExpressionByIdentifierHierarchy`.
 */
export function buildMemberExpressionByIdentifierHierarchy(
  hierarchy: Array<string | t.Identifier | t.ThisExpression>,
) {
  const normalizedHierarchy = hierarchy.map(
    path => (typeof path === 'string' ? t.identifier(path) : path),
  );

  let currentMemberExpression: t.MemberExpression;

  for (let i = 1; i <= normalizedHierarchy.length - 1; i++) {
    const left = normalizedHierarchy[i - 1];
    const right = normalizedHierarchy[i];
    if (left && right) {
      currentMemberExpression = t.memberExpression(
        i === 1 ? left : currentMemberExpression!,
        right,
      );
    }
  }

  return currentMemberExpression!;
}
