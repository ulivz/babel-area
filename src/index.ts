/**
 * Module dependencies
 */
import * as t from '@babel/types';
import { parse } from '@babel/parser';
import generate from '@babel/generator';
import traverse, { VisitNode, NodePath, Binding } from '@babel/traverse';
import * as helpers from './helpers';
import * as types from './types';

export * from './transform';

export {
  t,
  helpers,
  types,
  parse,
  traverse,
  generate,
  VisitNode,
  NodePath,
  Binding,
};
