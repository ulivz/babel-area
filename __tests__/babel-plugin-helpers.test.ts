/**
 * Module dependencies
 */
import { t, helpers } from '../src';
import { TestUtils } from './test-utils';

const args = [
  `
const foo = 'bar';
`,
  {
    start: {
      line: 2,
      column: 5,
    },
    end: {
      line: 2,
      column: 7,
    },
  },
  'error',
] as const;

it('getHighlightCodeString', () => {
  expect(
    TestUtils.stripAnsi(
      helpers.getHighlightCodeString(...args),
    ),
  ).toMatchSnapshot();
});

it('HighlightCodeError', () => {
  const error = new helpers.HighlightCodeError(...args);
  expect(
    TestUtils.stripAnsi(error.message),
  ).toMatchSnapshot();
});

describe('buildMemberExpressionByIdentifierHierarchy', () => {
  it('string - a.b', () => {
    expect(
      TestUtils.renderBabelNode(helpers.buildMemberExpressionByIdentifierHierarchy(['a', 'b'])),
    ).toBe('a.b;');
  });

  it('string - a.b.c', () => {
    expect(
      TestUtils.renderBabelNode(helpers.buildMemberExpressionByIdentifierHierarchy(['a', 'b', 'c'])),
    ).toBe('a.b.c;');
  });

  it('identifier - a.b', () => {
    expect(
      TestUtils.renderBabelNode(helpers.buildMemberExpressionByIdentifierHierarchy([
        t.identifier('a'),
        t.identifier('b'),
      ])),
    ).toBe('a.b;');
  });

  it('identifier - a.b', () => {
    expect(
      TestUtils.renderBabelNode(helpers.buildMemberExpressionByIdentifierHierarchy([
        t.identifier('a'),
        t.identifier('b'),
        t.identifier('c'),
      ])),
    ).toBe('a.b.c;');
  });
});
