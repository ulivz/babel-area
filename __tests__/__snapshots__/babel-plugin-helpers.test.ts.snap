// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightCodeError 1`] = `
"  1 |
> 2 | const foo = 'bar';
    |       ^ error
  3 |"
`;

exports[`getHighlightCodeString 1`] = `
"  1 |
> 2 | const foo = 'bar';
    |      ^ error
  3 |"
`;
